File: 864-image-overlap/image-overlap.cpp
class Solution {
public:
    int largestOverlap(vector<vector<int>>& img1, vector<vector<int>>& img2) {
         int n = img1.size();
        unordered_map<int, int> mp;

        vector<pair<int,int>> A, B;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (img1[i][j] == 1) A.push_back({i, j});
                if (img2[i][j] == 1) B.push_back({i, j});
            }
        }

        int res = 0;

        for (auto &a : A) {
            for (auto &b : B) {
                int dx = a.first - b.first;
                int dy = a.second - b.second;

                int key = dx * 100 + dy;
                mp[key]++;
                res = max(res, mp[key]);
            }
        }
        return res;
    }
};

File: 864-image-overlap/image-overlap.java
class Solution {
    public int largestOverlap(int[][] img1, int[][] img2) {
    int n = img1.length;
        List<int[]> A = new ArrayList<>();
        List<int[]> B = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (img1[i][j] == 1) A.add(new int[]{i, j});
                if (img2[i][j] == 1) B.add(new int[]{i, j});
            }
        }

        Map<String, Integer> map = new HashMap<>();
        int max = 0;

        for (int[] a : A) {
            for (int[] b : B) {
                int dx = a[0] - b[0];
                int dy = a[1] - b[1];

                String key = dx + "," + dy;
                int count = map.getOrDefault(key, 0) + 1;
                map.put(key, count);

                max = Math.max(max, count);
            }
        }

        return max;     
    }
}